core/src/main/java/io/grpc/internal/InstantTimeProvider.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2024 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link InstantTimeProvider} resolves InstantTimeProvider which implements {@link TimeProvider}.
+ */
+final class InstantTimeProvider implements TimeProvider {
+  private Method now;
+  private Method getNano;
+  private Method getEpochSecond;
+
+  public InstantTimeProvider(Class<?> instantClass) throws NoSuchMethodException {
+    this.now = instantClass.getMethod("now");
+    this.getNano = instantClass.getMethod("getNano");
+    this.getEpochSecond = instantClass.getMethod("getEpochSecond");
+  }
+
+  @Override
+  public long currentTimeNanos() {
+    try {
+      Object instant = now.invoke(null);
+      int nanos = (int) getNano.invoke(instant);
+      long epochSeconds = (long) getEpochSecond.invoke(instant);
+      return TimeUnit.SECONDS.toNanos(epochSeconds) + nanos;
```

Comment:
If the return value of TimeUnit.SECONDS.toNanos(epochSeconds) is Long.MAX_VALUE then just return it without adding nanos. No need to check for MIN_VALUE as Eric mentioned.

Reply:
Even if we do a check for long.MAX_VALUE the probability of meeting "if condition" being TRUE is exactly once and probability of invoking this method in that point in time is negligible.

Referring to Eric's below comment, he is also in the agreement to live with it.

![image](https://github.com/user-attachments/assets/acc37b49-b004-476f-89d8-307a8354a791)

Reply:
It doesn't change the discussion, but I do want to make clear: The probability _isn't_ exactly once, because TimeUnit.toNanos saturates at MAX_VALUE. It is, "every result of toNanos after some point in 2262 will be MAX_VALUE."

And not that it matters, but I'll take this opportunity to inform y'all that [LongMath.saturateAdd()](https://javadoc.io/doc/com.google.guava/guava/latest/com/google/common/math/LongMath.html#saturatedAdd(long,long)) exists and is convenient when you're doing lots of this sort of thing.

Reply:
Fixed